pep-0591.rst

```diff
@@ -0,0 +1,293 @@
+PEP: 591
+Title: Adding a final qualifier to typing
+Author: Michael J. Sullivan <sully@msully.net>, Ivan Levkivskyi <levkivskyi@gmail.com>
+Discussions-To: typing-sig@python.org
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 15-Mar-2019
+Post-History:
+
+
+Abstract
+========
+
+This PEP proposes a "final" qualifier to be added to the ``typing``
+module---in the form of a ``final`` decorator and a ``Final`` type
+annotation---to serve three related purposes:
+
+* Declaring that a method should not be overridden
+* Declaring that a class should not be subclassed
+* Declaring that a variable or attribute should not be reassigned
+
+
+Motivation
+==========
+
+The ``final`` decorator
+-----------------------
+The current ``typing`` module lacks a way to restrict the use of
+inheritance or overriding at a typechecker level. This is a common
+feature in other object-oriented languages (such as Java), and is
+useful for reducing the potential space of behaviors of a class,
+easing reasoning.
+
+Some situations where a final class or method may be useful include:
+
+* A class wasn’t designed to be subclassed or a method wasn't designed
+  to be overridden. Perhaps it would not work as expected, or be
+  error-prone.
+* Subclassing or overriding would make code harder to understand or
+  maintain. For example, you may want to prevent unnecessarily tight
+  coupling between base classes and subclasses.
+* You want to retain the freedom to arbitrarily change the class
+  implementation in the future, and these changes might break
+  subclasses.
+
+The ``Final`` annotation
+------------------------
+
+The current ``typing`` module lacks a way to indicate that a variable
+will not be assigned to. This is a useful feature in several
+situations:
+
+* Preventing unintended modification of module and class level
+  constants and documenting them as constants in a checkable way.
+* Creating a read-only attribute that may not be overridden by
+  subclasses. (``@property`` can make an attribute read-only but
+  does not prevent overriding)
+* Allowing a name to be used in situations where ordinarily a literal
+  is expected (for example as a file name for ``NamedTuple``, a tuple
```

Comment:
Sorry I'm late to the party, but I'm guessing "file name" is a typo for "field name".